central/clusterinit/store/store.go

```diff
@@ -145,3 +166,108 @@ func (w *storeImpl) Revoke(ctx context.Context, id string) error {
 	}
 	return nil
 }
+
+// InitiateClusterRegistration checks if another registration using the CRS with the provided CRS ID is possible.
+// If the provided id belongs to an init bundle, then registrations is always allowed, without any bookkeeping.
+func (w *storeImpl) InitiateClusterRegistration(ctx context.Context, initArtifactId, clusterName string) error {
+	w.uniqueUpdateMutex.Lock()
+	defer w.uniqueUpdateMutex.Unlock()
+
+	log.Infof("Attempting registration for cluster %s using init artifact %s", clusterName, initArtifactId)
+	initArtifactMeta, err := w.Get(ctx, initArtifactId)
+	if err != nil {
+		return errors.Wrapf(err, "retrieving init artifact meta data for ID %q", initArtifactId)
+	}
+
+	if initArtifactMeta.GetIsRevoked() {
+		log.Warnf("Init artifact %s is revoked, registration of cluster %s not allowed", initArtifactId, clusterName)
+	}
+
+	switch initArtifactMeta.Version {
+	case storage.InitBundleMeta_CRS:
+	case storage.InitBundleMeta_INIT_BUNDLE:
+		return nil
+	}
+
+	crsMeta := initArtifactMeta
+	maxRegistrations := crsMeta.MaxRegistrations
+	registrationsInitiatedSet := set.NewStringSet(crsMeta.RegistrationsInitiated...)
+	registrationsCompletedSet := set.NewStringSet(crsMeta.RegistrationsCompleted...)
+	numRegistrationsTotal := uint64(len(registrationsInitiatedSet) + len(registrationsCompletedSet))
+	if maxRegistrations > 0 && numRegistrationsTotal >= maxRegistrations {
```

Comment:
Hm, perhaps we should check `maxRegistrations > 0` earlier and skip the bookkeeping otherwise?
Otherwise the set(s) can grow without bound, right?
I'm specifically worried about a use case where the user has a long-lived CRS and creates ephemeral clusters often (in CI or something).
If you agree then obviously `MarkClusterRegistrationComplete` needs changes accordingly.